core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/Histogram.java

```diff
@@ -103,13 +107,27 @@ public int compare(Histogram.Bucket b1, Histogram.Bucket b2) {
             }
         });
 
+        public static final OrderSpec SPEC = new OrderSpec() {
```

Comment:
I created this interface to help centralize the code that depends on aggregation specific orders. e.g. `histogram` aggs support a `_key` or `_time` order and `terms` aggs support a `_term` order. The serialization logic (`byte id` values) was different and the `CompoundOrder` sometimes needs to automatically add a "tie-breaker to avoid non-deterministic ordering if all user provided comparators return 0." Which is `_key` ascending for `histogram` aggs and `_term` ascending for `terms` aggs.

`OrderSpec#getOrders()` returns all of the aggregation specific sorts so `InternalOrder.Streams#readOrder` can deserialize a `byte id` back into the correct `BucketOrder`. Note aggregation and compound order serialization was the same for `terms` and `histogram` so I moved that code into that `readOrder` method.

`OrderSpec#getAbsoluteOrder` returns the "tie-breaker" `BucketOrder` needed in `CompoundOrder`

Each agg creates its own `OrderSpec` instance that is required when creating a `CompoundOrder` or deserializing a `BucketOrder`.

Some thoughts:
I can change `OrderSpec#getOrders()` into something like `OrderSpec#getOrder(byte id)` if that makes more sense.
Another solution may be to serialize the `key` and `asc` values instead of just an ID.